File: desktop/app/scripts/import_asp_history_test.py
"""The one-time desktop import preserves old ASP events without a receipt."""

import importlib.util
import sqlite3
import sys
import tempfile
from pathlib import Path


def main() -> None:
    script = Path(__file__).resolve().parent / "import-asp-history.py"
    spec = importlib.util.spec_from_file_location("import_asp_history", script)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"could not load {script}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    def fetch(_base, path, _token=None):
        if path == "/operator":
            return {"token": "secret"}
        if path == "/conversations":
            return {"conversations": [{"conversation_id": "sess_1", "kind": "group", "agents": []}]}
        if path == "/sessions/sess_1/events":
            return {
                "events": [
                    {
                        "event_id": "evt_1",
                        "sequence": 0,
                        "type": "session.message",
                        "payload": {"content": "hello"},
                    }
                ]
            }
        raise AssertionError(path)

    module.fetch = fetch
    with tempfile.TemporaryDirectory() as tmp:
        database = Path(tmp) / "chat.sqlite"
        sys.argv = [str(script), "--database", str(database)]
        module.main()
        with sqlite3.connect(database) as connection:
            assert connection.execute("SELECT count(*) FROM conversations").fetchone()[0] == 1
            assert connection.execute("SELECT count(*) FROM events").fetchone()[0] == 1
            assert connection.execute("SELECT count(*) FROM acknowledgements").fetchone()[0] == 0
    print("ok")


if __name__ == "__main__":
    main()

File: desktop/app/test/import-asp-history.test.mjs
import assert from "node:assert/strict";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import test from "node:test";
import { openChatStore } from "../main/chat-store.mjs";
import { importAspHistory } from "../scripts/import-asp-history.mjs";

test("legacy ASP import preserves events without acknowledging them", async (t) => {
  const root = mkdtempSync(path.join(tmpdir(), "magi-history-import-"));
  t.after(() => rmSync(root, { recursive: true, force: true, maxRetries: 5, retryDelay: 100 }));
  const database = path.join(root, "chat.sqlite");
  const fetcher = async (url, options) => {
    const endpoint = new URL(url).pathname;
    if (endpoint === "/operator") return Response.json({ token: "secret" });
    assert.equal(options.headers.Authorization, "Bearer secret");
    if (endpoint === "/conversations") return Response.json({ conversations: [{ conversation_id: "sess_1", kind: "group" }] });
    if (endpoint === "/sessions/sess_1/events") return Response.json({ events: [{ event_id: "evt_1", sequence: 0, type: "session.message", payload: { content: "hello" } }] });
    throw new Error(`unexpected request: ${endpoint}`);
  };
  assert.deepEqual(await importAspHistory({ database, fetcher }), { conversations: 1, events: 1, database });
  const store = await openChatStore(database);
  try {
    assert.equal(store.listConversations().length, 1);
    assert.equal(store.listEvents("sess_1").length, 1);
    assert.equal(store.pendingAcks("sess_1").length, 1);
  } finally { store.close(); }
});
